go/samples/rag/main.go

```diff
@@ -0,0 +1,151 @@
+// Start the server listening on port 3100:
+//
+//	go run . &
+//
+// Tell it to run a flow:
+//
+//	curl -d '{"key":"/flow/simpleQaFlow/simpleQaFlow", "input":{"start": {"input":{"question": "What is the capital of UK?"}}}}' http://localhost:3100/api/runAction
+//
+// In production mode (GENKIT_ENV missing or set to "prod"):
+// Start the server listening on port 3400:
+//
+//	go run . &
+//
+// Tell it to run a flow:
+//
+//   curl -d '{"question": "What is the capital of UK?"}' http://localhost:3400/simpleQaFlow
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/firebase/genkit/go/genkit"
+	"github.com/firebase/genkit/go/plugins/dotprompt"
+	"github.com/firebase/genkit/go/plugins/googleai"
+	"github.com/firebase/genkit/go/plugins/localvec"
+	"github.com/invopop/jsonschema"
+)
+
+const simpleQaPromptTemplate = `
+You're a helpful agent that answers the user's common questions based on the context provided.
+
+Here is the user's query: {{query}}
+
+Here is the context you should use: {{context}}
+
+Please provide the best answer you can.
+`
+
+type simpleQaInput struct {
+	Question string `json:"question"`
+}
+
+type simpleQaPromptInput struct {
+	Query   string `json:"query"`
+	Context string `json:"context"`
+}
+
+func main() {
+	apiKey := os.Getenv("GOOGLE_GENAI_API_KEY")
+	if apiKey == "" {
+		fmt.Fprintln(os.Stderr, "coffee-shop example requires setting GOOGLE_GENAI_API_KEY in the environment.")
+		fmt.Fprintln(os.Stderr, "You can get an API key at https://ai.google.dev.")
+		os.Exit(1)
+	}
+
+	if err := googleai.Init(context.Background(), "gemini-1.0-pro", apiKey); err != nil {
+		log.Fatal(err)
+	}
+
+	simpleQaPrompt, err := dotprompt.Define("simpleQaPrompt",
+		simpleQaPromptTemplate,
+		&dotprompt.Config{
+			Model:        "google-genai/gemini-1.0-pro",
+			InputSchema:  jsonschema.Reflect(simpleQaPromptInput{}),
+			OutputFormat: ai.OutputFormatText,
+		},
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	embedder, err := googleai.NewEmbedder(context.Background(), "embedding-001", apiKey)
+	if err != nil {
+		log.Fatal(err)
+	}
+	localDb, err := localvec.New(context.Background(), "/tmp/", "simpleQa", embedder, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	genkit.DefineFlow("simpleQaFlow", func(ctx context.Context, input *simpleQaInput, _ genkit.NoStream) (string, error) {
+		d1 := ai.DocumentFromText("Paris is the capital of France", nil)
+		d2 := ai.DocumentFromText("USA is the largest importer of coffee", nil)
+		d3 := ai.DocumentFromText("Water exists in 3 states - solid, liquid and gas", nil)
+
+		indexerReq := &ai.IndexerRequest{
+			Documents: []*ai.Document{d1, d2, d3},
+		}
+		localDb.Index(ctx, indexerReq)
+
+		dRequest := ai.DocumentFromText(input.Question, nil)
+		retrieverReq := &ai.RetrieverRequest{
+			Document: dRequest,
+		}
+		response, err := localDb.Retrieve(ctx, retrieverReq)
+		if err != nil {
+			return "", err
+		}
+
+		var context string
```

Comment:
Better coding would be a `strings.Builder`.

```Go
    var sb strings.Builder
    for _, d := range response.Documents {
        sb.WriteString(d.Content[0].Text())
        sb.WriteByte('\n')
    }

    ...

        Context: sb.String(),

Reply:
Done.